safehttp/form.go

```diff
@@ -0,0 +1,219 @@
+// limitations under the License.
+
+package safehttp
+
+import (
+	"errors"
+	"mime/multipart"
+	"strconv"
+)
+
+// Form TODO
+type Form struct {
+	values map[string][]string
+	err    error
+	parsed bool
+}
+
+// MultipartForm TODO
+type MultipartForm struct {
+	form Form
+	file map[string][]*multipart.FileHeader
+}
+
+// Int TODO
+func (f *Form) Int(paramName string, defaultValue int) int {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.Atoi(vals[0])
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Uint TODO
+func (f *Form) Uint(paramName string, defaultValue uint64) uint64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseUint(vals[0], 10, 0)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+func (f *Form) String(paramName string, defaultValue string) string {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	return vals[0]
+}
+
+// Float64 TODO
+func (f *Form) Float64(paramName string, defaultValue float64) float64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseFloat(vals[0], 64)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Bool TODO
+func (f *Form) Bool(paramName string, defaultValue bool) bool {
+	if f.err != nil {
+		return defaultValue
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	if vals[0] != "true" {
+		if vals[0] != "false" {
+			f.err = errors.New("values of form parameter " + paramName + " not a boolean")
+		}
+		return false
+	}
+	return true
+}
+
+// Slice TODO
+func (f *Form) Slice(slice interface{}, paramName string) {
+	if f.err != nil {
+		slice = nil
+		return
+	}
+	if !f.parsed {
+		panic("form has not been parsed")
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		slice = nil
+	}
+	len := len(vals)
+	switch data := slice.(type) {
+	case *[]string:
+		res := *data
+		res = make([]string, 0, len)
+		for _, x := range vals {
+			res = append(res, x)
```

Comment:
You need to write some tests for this, because my hunch is that this code doesn't work as you expected. I'm also fine if you add `Slice` in a separate PR if you want to make this a quick one. :)

Reply:
Partly resolved in [7b2f76b](url), more tests to come